libs/ledger/src/protocols/notarisation_service.cpp

```diff
@@ -0,0 +1,392 @@
+  {
+    return {};
+  }
+  return notarisations_being_built_.at(height);
+}
+
+void NotarisationService::NotariseBlock(BlockBody const &block)
+{
+  FETCH_LOCK(mutex_);
+
+  // Return if not eligible to notarise, block is too far in the past of the highest notarised block
+  // or beyond the window of this aeon
+  if (!active_exe_unit_ || block.block_number < active_exe_unit_->aeon.round_start ||
+      block.block_number >= active_exe_unit_->aeon.round_end)
+  {
+    return;
+  }
+  // Return if block has already been notarised
+  if (notarisations_built_.find(block.block_number) != notarisations_built_.end() &&
+      notarisations_built_.at(block.block_number).find(block.hash) !=
+          notarisations_built_.at(block.block_number).end())
+  {
+    // TODO(JMW): Block has already been notarised -> tell main chain
+    return;
+  }
+  // Return if block is too far in the past of the head of the notarised chain
+  if (block.block_number < BlockNumberCutoff())
+  {
+    return;
+  }
+
+  // If first block of the aeon then qual should match
+  if (block.block_number == active_exe_unit_->aeon.round_start)
+  {
+    assert(block.block_entropy.qualified == active_exe_unit_->manager.qual());
+  }
+
+  // Determine rank of miner in qual
+  auto entropy_ranked_cabinet = Consensus::QualWeightedByEntropy(
+      active_exe_unit_->manager.qual(), block.block_entropy.EntropyAsU64());
+  auto miner_position =
+      std::find(entropy_ranked_cabinet.begin(), entropy_ranked_cabinet.end(), block.miner_id);
+  assert(miner_position != entropy_ranked_cabinet.end());
+  auto miner_rank =
+      static_cast<uint32_t>(std::distance(entropy_ranked_cabinet.begin(), miner_position));
+
+  // Check if we have previously signed a higher ranked block at the same height
+  if (previous_notarisation_rank_.find(block.block_number) != previous_notarisation_rank_.end())
+  {
+    if (previous_notarisation_rank_.at(block.block_number) > miner_rank)
+    {
+      return;
+    }
+  }
+
+  // Sign and verify own notarisation and then save for peers to query
+  auto notarisation = active_exe_unit_->manager.Sign(block.hash);
+  assert(active_exe_unit_->manager.VerifySignatureShare(block.hash, notarisation.signature,
+                                                        notarisation.identity.identifier()));
+  // Sign notarisation with ecdsa private key
+  auto ecdsa_sig = certificate_->Sign(
+      (serializers::MsgPackSerializer() << block.hash << notarisation.signature).data());
+  notarisations_being_built_[block.block_number][block.hash].insert(
+      {notarisation.identity.identifier(), SignedNotarisation(ecdsa_sig, notarisation.signature)});
+
+  // Set highest notarised block rank for this block height
+  previous_notarisation_rank_[block.block_number] = miner_rank;
+}
+
+std::vector<std::weak_ptr<core::Runnable>> NotarisationService::GetWeakRunnables()
+{
+  return {state_machine_};
+}
+
+void NotarisationService::NewAeonExeUnit(SharedAeonExecutionUnit const &keys)
```

Comment:
Might be worth thinking about doing a full copy of this due to thread safety. It looks safe to me currently, but if you were to do 'setmessage' in this class it would be unsafe to also be generating entropy